File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
       int n = nums.length;
       int left = 1;
       int right = 1;
       int max = nums[0];
       for(int i=0; i<nums.length; i++){
            if(left==0){
                left=1;
            }
            if(right==0){
                right=1;
            }
            left*=nums[i];
            right*=nums[n-1-i];
            max = Math.max(Math.max(left, right),max);
       }return max;
    }
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        res = nums[0]
        curMin = 1
        curMax = 1
        for num in nums:
            tmp = curMin
            curMin = min(num, curMax * num, curMin * num)
            curMax = max(num, curMax * num, tmp * num)
            res = max(res, curMax)
        return res